### PR TITLE
Issue #34: added "Lookup URI" button for URI text fields.

### PR DESCRIPTION
**Pull request by kodecharlie**
Issue #34.  See my comments there.

project.xml is a NetBeans config file.

### CODE REVIEW

File: nbproject/project.xml [hunk 1]
```diff
@@ -0,0 +1,64 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (aclarkxyz) [inline nbproject/project.xml]: This file has to go; Eclipse only.

File: src/com/cdd/bao/editor/DetailPane.java [hunk 2]
```diff
@@ -75,7 +68,8 @@
 	{
 		Lineup line;
 		Schema.Value sourceVal;
-		TextField fieldURI, fieldName;
+		URIRowLine fieldURI;
```
Comment (aclarkxyz) [inline src/com/cdd/bao/editor/DetailPane.java]: Multiple declarations on the same line are allowed (though not required), so roll back the change.

File: src/com/cdd/bao/editor/DetailPane.java [hunk 3]
```diff
@@ -335,19 +329,43 @@ public void actionValueMove(int dir)
 
 		valueList.get(newIndex).fieldURI.requestFocus();
 	}
-	public void actionLookupURI()
+
+	@Override
+	public void actionLookupURI(int focusIndex)
 	{
-		if (focusIndex < 0) return;
+		System.out.println("focusIndex=" + focusIndex + "; valueList.length=" + valueList.size());
```
Comment (aclarkxyz) [inline src/com/cdd/bao/editor/DetailPane.java]: Remove.

File: src/com/cdd/bao/editor/DetailPane.java [hunk 4]
```diff
@@ -335,19 +329,43 @@ public void actionValueMove(int dir)
...
 		ValueWidgets vw = valueList.get(focusIndex);
 		String uri = ModelSchema.expandPrefix(vw.fieldURI.getText());
-		if (uri.length() == 0) {actionLookupName(); return;}
-	
+		if (uri.length() == 0)
```
Comment (aclarkxyz) [inline src/com/cdd/bao/editor/DetailPane.java]: This isn't wrong, but multiple statements on the same line are allowed; by default leave it (i.e. roll it back).

File: src/com/cdd/bao/editor/DetailPane.java [hunk 5]
```diff
@@ -335,19 +329,43 @@ public void actionValueMove(int dir)
...
+		}
+
 		Vocabulary vocab = Vocabulary.globalInstance();
-		
+
 		String label = vocab.getLabel(uri), descr = vocab.getDescr(uri);
-		if (label == null) {actionLookupName(); return;}
-		if (vw.fieldName.getText().length() == 0) vw.fieldName.setText(label);
-		if (descr != null && vw.fieldDescr.getText().length() == 0) vw.fieldDescr.setText(descr);
+		if (label == null)
```
Comment (aclarkxyz) [inline src/com/cdd/bao/editor/DetailPane.java]: Too wordy; roll it back.

File: src/com/cdd/bao/editor/DetailPane.java (hunk 2 above)
Comment (gedeck) [reply src/com/cdd/bao/editor/DetailPane.java]: Type has changed

File: src/com/cdd/bao/util/URIRowLine.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import javafx.scene.control.TextField;
+import javafx.scene.control.Tooltip;
+
+public class URIRowLine extends RowLine
+{
+	private URITextFieldDelegate delegate;
+
+	private TextField getURITextField()
+	{
+		for (Node n : getChildren())
```
Comment (aclarkxyz) [inline src/com/cdd/bao/util/URIRowLine.java]: Make it more concise. Use on-same-line for things that are really short, and when there's nothing complicated going on that requires extra lines.

File: src/com/cdd/bao/util/URIRowLine.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+		{
+			if (n instanceof TextField)
+				return (TextField) n;
```
Comment (aclarkxyz) [inline src/com/cdd/bao/util/URIRowLine.java]: No space after cast, i.e. (TextField)n

File: src/com/cdd/bao/util/URIRowLine.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+		if (focusIndex >= 0)
+		{
+			lookupUriBtn.setOnAction(event -> this.delegate.actionLookupURI(focusIndex));
+		}
+
+		this.add(tfURI, 1);
+		this.add(lookupUriBtn, 0);
+	}
+
+	/**
```
Comment (aclarkxyz) [inline src/com/cdd/bao/util/URIRowLine.java]: Javadoc-style comments aren't really being used in this project, so just use concise descriptions, with "// {what it does}", especially for methods that are as obvious as this.

File: src/com/cdd/bao/util/URITextFieldDelegate.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
+/*
```
Comment (aclarkxyz) [inline src/com/cdd/bao/util/URITextFieldDelegate.java]: The functionality in this class isn't enough to justify having its own separate file; make it into an inner class instead.

File: src/com/cdd/bao/editor/DetailPane.java (hunk 2 above)
Comment (aclarkxyz) [reply src/com/cdd/bao/editor/DetailPane.java]: Right.

File: nbproject/project.xml (hunk 1 above)
Comment (kodecharlie) [reply nbproject/project.xml]: Will move to Eclipse today.

File: src/com/cdd/bao/editor/DetailPane.java [hunk 10]
```diff
@@ -874,7 +879,7 @@ private void pressedAnnotationButton(Schema.Assignment assn, int idx)
 	}
 
 	// respond to focus so that one of the blocks gets a highlight
-	private void observeFocus(Control field, final int idx)
+	public void observeFocus(Control field, final int idx)
```
Comment (gedeck) [inline src/com/cdd/bao/editor/DetailPane.java]: Out of curiosity, what is the reason to change the visibility from `private` to `public`?

Comment (gedeck) [reply src/com/cdd/bao/editor/DetailPane.java]: Ignore that question

File: src/com/cdd/bao/util/URITextFieldDelegate.java (hunk 9 above)
Comment (kodecharlie) [reply src/com/cdd/bao/util/URITextFieldDelegate.java]: Its an interface.

Comment (aclarkxyz) [reply src/com/cdd/bao/util/URITextFieldDelegate.java]: That's fine: interfaces can be inner classes. (Making a separate class for line + button widget is already a bit of a stretch, but adding 2 new files is definitively over the line.)

File: src/com/cdd/bao/util/URIRowLine.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+import javafx.scene.control.TextField;
+import javafx.scene.control.Tooltip;
+
+public class URIRowLine extends RowLine
+{
+	private URITextFieldDelegate delegate;
+
+	private TextField getURITextField()
+	{
+		for (Node n : getChildren())
```
Comment (aclarkxyz) [inline src/com/cdd/bao/util/URIRowLine.java]: The variable name "n" is pretty much exclusively used for integer enumeration, so use something longer, like "node".
